src/xmipp/libraries/reconstruction_starpu/reconstruct_fourier_starpu_util.h

```diff
@@ -0,0 +1,430 @@
+					newPos[0] = maxVolumeIndexYZ - x;
+					newPos[1] = maxVolumeIndexYZ - y;
+					newPos[2] = maxVolumeIndexYZ - z;
+					output[newPos[2]][newPos[1]][newPos[0]-maxVolumeIndexX] += f(input[z][y][x]);
+				} else {
+					// copy with X shifted by (-halfSize)
+					output[z][y][x-maxVolumeIndexX] += input[z][y][x];
+				}
+			}
+		}
+	}
+	// free original data
+	release(input, maxVolumeIndexYZ+1, maxVolumeIndexYZ+1);
+	// set new data
+	input = output;
+}
+
+/**
+ * Method will take temp spaces (containing complex conjugate values
+ * in the 'right X side'), transfer them to 'left X side' and remove
+ * the 'right X side'. As a result, the X dimension of the temp spaces
+ * will be half of the original.
+ */
+static void mirrorAndCropTempSpaces(std::complex<float>***& tempVolume, float***& tempWeights, const uint32_t maxVolumeIndexX, const uint32_t maxVolumeIndexYZ) {
+	mirrorAndCrop(tempWeights, &identity<float>, maxVolumeIndexX, maxVolumeIndexYZ);
+	mirrorAndCrop(tempVolume, &conjugate, maxVolumeIndexX, maxVolumeIndexYZ);
+}
+
+/**
+ * Method will enforce Hermitian symmetry, i.e will make sure
+ * that the values in temporal space at X=0 are complex conjugate of
+ * in respect to center of the space
+ */
+static void forceHermitianSymmetry(std::complex<float>***& tempVolume, float***& tempWeights, int maxVolumeIndexYZ) {
+	int x = 0;
+	for (int z = 0; z <= maxVolumeIndexYZ; z++) {
+		for (int y = 0; y <= maxVolumeIndexYZ/2; y++) {
+			int newPos[3];
+			// mirror against center of the volume, e.g. [0,0,0]->[size,size,size]. It will fit as the input space is one voxel biger
+			newPos[0] = x;
+			newPos[1] = maxVolumeIndexYZ - y;
+			newPos[2] = maxVolumeIndexYZ - z;
+			std::complex<float> tmp1 = 0.5f * (tempVolume[newPos[2]][newPos[1]][newPos[0]] + conj(tempVolume[z][y][x]));
+			float tmp2 = 0.5f * (tempWeights[newPos[2]][newPos[1]][newPos[0]] + tempWeights[z][y][x]);
+
+			tempVolume[newPos[2]][newPos[1]][newPos[0]] = tmp1;
+			tempVolume[z][y][x] = conj(tmp1);
+			tempWeights[newPos[2]][newPos[1]][newPos[0]] = tempWeights[z][y][x] = tmp2;
+		}
+	}
+}
+
+/**
+ * Method will in effect do the point-wise division of
+ * tempVolume and tempWeights
+ * (i.e. correct Fourier coefficients by proper weight)
+ */
+static void processWeights(std::complex<float>***& tempVolume, float***& tempWeights, int maxVolumeIndexX, int maxVolumeIndexYZ, double paddingFactorProj, double paddingFactorVol, int imgSize) {
+	// Get a first approximation of the reconstruction
+	float corr2D_3D = static_cast<float>(pow(paddingFactorProj, 2.0) / (imgSize * pow(paddingFactorVol, 3.0)));
+	for (int z = 0; z <= maxVolumeIndexYZ; z++) {
+		for (int y = 0; y <= maxVolumeIndexYZ; y++) {
+			for (int x = 0; x <= maxVolumeIndexX; x++) {
+				float weight = tempWeights[z][y][x];
+				if (fabs(weight) > 1e-3) {
+					weight = 1.f/weight;
+				}
+
+				if (1.0/weight > ACCURACY)
+					tempVolume[z][y][x] *= corr2D_3D * weight;
+				else
+					tempVolume[z][y][x] = 0;
```

Comment:
This logic seems strange. Since `ACCURACY` is `1e-3`, this code will always do one of these things:
1. If `tempWeigths[][][]` is > `ACCURACY`: `tempVolume[z][y][x] *= corr2D_3D / tempWeigths[][][]`
2. If `tempWeigths[][][]` is < `ACCURACY`: `tempVolume[z][y][x] *= corr2D_3D * tempWeigths[][][]`
3. `tempVolume[z][y][x] = 0` in rare cases when `1/(1/weight) != weight` in just the correct way.

I am not sure what the mathematics behind this is, but it seems wrong. Perhaps the intended logic was something like:
```
if (weight > ACCURACY)
    tempVolume[z][y][x] *= corr2D_3D / weight;
else
    tempVolume[z][y][x] = 0;
```
@DStrelak ?
(Same logic is in both (CPU/CUDA) previous versions.)

Reply:
Opened https://github.com/I2PC/xmipp/pull/243 to fix it in other `reconstruct_fourier` variants.